### ISSUE TITLE
[Feature request] Conditional Random Fields support

### ISSUE THREAD

**Issue by bratao** [explosion/thinc#282]
Hello, 
I love thinc, is so much faster than anything else. But I really miss a CRF implementation in thinc.

@honnibal explained that Spacy NER do not use it, because: "If the above sounds strange to you, the main motivation is that we get to condition on structures built over arbitrarily far back transitions. This is especially helpful for the parser, but it's good for the NER too: our features include tokens for the first and last word of the currently open entity, which is not something you could express easily with a limited-order CRF. In order to make these features meaningful, the parser has to visit a realistic sample of transition sequences during training, including sequences that result from previous transition errors."

But for my own NER tasks( extracting information from a semi-structured document) , CRF really improve the results. In this rewrite, there is any chance of appearing a CRF implementation?

**Comment by honnibal** [comment]
Glad to hear the library looks good to you! A quick clarification though: Thinc's own implementations will often be slower than the same thing written in other frameworks, simply because there's so many years of engineering effort in e.g. PyTorch. The reason spaCy's models run faster is more down to architectural choices in the network.

You'll be able to use CRF models easily by wrapping other frameworks. We do actually have a beam implementation as well, which I use for a CRF-like model in spaCy's parser and NER...But the beam doesn't improve accuracy currently. 

I would suggest using the PyTorch wrapper for making a CRF. Soon I also hope we can support DyNet, which is especially efficient for small CRF models.

**Comment by naveenjafer** [comment]
@honnibal I am experimenting on NP Chunking using thinc. The SOTA I am using as referring uses CRF along with a BiLSTM. My initial objective was to use the implementation to build an alternate of the static rule-based NP chunker in SpaCy. The added advantage is that it would help us with Verb Phrases too.   

If this were to be implemented with SpaCy, then this would have to be a pure implementation with thinc and not involve wrappers of tensorflow or pytorch, in that case, does it make sense to experiment on a CRF layer for thinc too?
